internal/edittree/edittree.go

```diff
@@ -0,0 +1,1161 @@
+	e.setChildKey(keyHash, key)
+	if isComposite(value) {
+		e.setChildCompositeValue(keyHash, child)
+	} else {
+		e.setChildScalarValue(keyHash, value)
+	}
+	return child
+}
+
+func (e *EditTree) unsafeInsertArray(idx int, value *ast.Term) *EditTree {
+	child := NewEditTree(value)
+	// Collect insertion indexes above the insertion site for rewriting.
+	rewritesScalars := []int{}
+	rewritesComposites := []int{}
+	for i := idx; i < e.insertions.Length(); i++ {
+		if e.insertions.Element(i) == 1 {
+			if _, ok := e.childScalarValues[i]; ok {
+				rewritesScalars = append(rewritesScalars, i)
+				continue
+			}
+			if _, ok := e.childCompositeValues[i]; ok {
+				rewritesComposites = append(rewritesComposites, i)
+				continue
+			}
+			panic(fmt.Errorf("invalid index %d during Insert operation", i))
+		}
+	}
+	// Do rewrites in reverse order to make room for the newly-inserted element.
+	for i := len(rewritesScalars) - 1; i >= 0; i-- {
+		originalIdx := rewritesScalars[i]
+		rewriteIdx := rewritesScalars[i] + 1
+		v := e.childScalarValues[originalIdx]
+		e.deleteChildValue(originalIdx)
+		e.setChildScalarValue(rewriteIdx, v)
+	}
+	for i := len(rewritesComposites) - 1; i >= 0; i-- {
+		originalIdx := rewritesComposites[i]
+		rewriteIdx := rewritesComposites[i] + 1
+		v := e.childCompositeValues[originalIdx]
+		e.deleteChildValue(originalIdx)
+		e.setChildCompositeValue(rewriteIdx, v)
+	}
+	// Insert new element in children array, bump bit-vector over by 1.
+	if idx == e.insertions.Length() {
+		e.insertions.Append(1)
+	} else {
+		e.insertions.Insert(1, idx)
+	}
+	if isComposite(value) {
+		e.setChildCompositeValue(idx, child)
+	} else {
+		e.setChildScalarValue(idx, value)
+	}
+	return child
+}
+
+// Delete removes a child of e, or else creates a delete node for a term
+// already present in e. It then returns the deleted child EditTree node.
+func (e *EditTree) Delete(key *ast.Term) (*EditTree, error) {
+	if e.value == nil {
+		return nil, fmt.Errorf("deleted node encountered during delete operation")
+	}
+	if key == nil {
+		return nil, fmt.Errorf("nil key provided for delete operation")
+	}
+
+	switch e.value.Value.(type) {
+	case ast.Object:
+		keyHash, found := e.getKeyHash(key)
+		// If child found, replace with delete node. If delete node already existed, error.
+		if found {
+			if child, ok := e.childScalarValues[keyHash]; ok {
+				if child == nil {
+					return nil, fmt.Errorf("cannot delete the already deleted scalar node for key %v", key)
+				}
+				e.setChildKey(keyHash, key)
+				e.setChildScalarValue(keyHash, nil)
+				return NewEditTree(child), nil
+			}
+			if child, ok := e.childCompositeValues[keyHash]; ok {
+				if child == nil {
+					return nil, fmt.Errorf("cannot delete the already deleted composite node for key %v", key)
+				}
+				e.setChildKey(keyHash, key)
+				e.setChildCompositeValue(keyHash, nil)
+				return child, nil
+			}
+			panic(fmt.Errorf("hash value %d not found in scalar or composite child maps", keyHash))
```

Comment:
Do we reach here? Should this be an error?

Reply:
Re: reachability? :: It should _theoretically_ be impossible to reach that panic, but only because the invariant (one map or the other will _always_ have the value) are carefully preserved elsewhere.

If we ever have to significantly modify the `EditTree` in the future, and the bookkeeping gets messed up, we'll get that specific panic, and know roughly what went wrong.

Re: should this be an error? :: I would say that it should remain a panic, because if we ever reach that case, then something is _very badly wrong_, such as the tree being in an illegal/inconsistent state. That kind of situation is not recoverable, and will break the lookups/bookkeeping elsewhere in the data structure.

Reply:
Ok. Then can you please add a comment about when such a scenario could happen and its consequences. Thanks.

Reply:
Done. Added comments to the 3x sites where that scenario could happen, with short explanations of how one could get there through mangling the data structure outside of the public methods.